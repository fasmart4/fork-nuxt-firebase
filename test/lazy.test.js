const { resolve } = require('path')
const fs = require('fs-extra')
const { Nuxt, Builder } = require('nuxt')
const FirebaseModule = require('..')

jest.mock('firebase/app', () => {
  const { getProxyMock } = require('proxy-mock-js')
  const spyFn = (name, fn) => jest.fn(fn)
  const session = getProxyMock({}, 'firebaseSession', spyFn)

  /* const session = new Proxy({}, {
    get(target, name) {
      return jest.fn(() => ({}))
    }
  }) */

  return {
    apps: [session],
    messaging: {
      isSupported: jest.fn()
    }
  }
})

describe('lazy', () => {
  let nuxt
  const buildDir = resolve(__dirname, '.nuxt-lazy')

  beforeAll(async () => {
    const config = {
      rootDir: resolve(__dirname, '..'),
      buildDir,
      srcDir: resolve(__dirname, 'fixture'),
      modules: [
        [FirebaseModule, {
          injectModule: false,
          lazy: true,
          config: {
            // REQUIRED: Official config for firebase.initializeApp(config):
            apiKey: '<apiKey>',
            authDomain: '<authDomain>',
            databaseURL: '<databaseURL>',
            projectId: '<projectId>',
            storageBucket: '<storageBucket>',
            messagingSenderId: '<messagingSenderId>',
            appId: '<appId>',
            measurementId: '<measurementId>'
          },
          services: {
            analytics: true,
            auth: true,
            firestore: true,
            functions: true,
            messaging: true,
            performance: true,
            database: true,
            remoteConfig: true,
            storage: true
          }
        }]
      ]
    }
    config.dev = false

    nuxt = new Nuxt(config)
    const BundleBuilder = { build: _ => _ }
    const builder = new Builder(nuxt, BundleBuilder)
    await builder.build()
  }, 60000)

  afterAll(async () => {
    await nuxt.close()
  })

  test('plugin contents', async () => {
    const content = await fs.readFile(resolve(buildDir, 'firebase/index.js'), { encoding: 'utf8' })

    expect(content).toContain('fire.auth = null')
    expect(content).toContain('fire.messaging = null')

    expect(content).not.toContain('Mess')

    expect(content).toMatchSnapshot()
  })

  test('exec plugin (server)', async () => {
    const Plugin = await import(resolve(buildDir, 'firebase/index.js')).then(m => m.default || m)
    const ctx = {}
    const inject = jest.fn()

    await Plugin(ctx, inject)
    expect(inject).toHaveBeenCalledTimes(1)
    expect(ctx.$fire).toBeDefined()
    expect(ctx.$fireModule).toBeUndefined()

    expect(ctx.$fire.appReady).toBeDefined()

    // TODO: test the appReady call. Doesnt work atm due to Jest not detecting
    // a async callback in the session var. Not fully sure whats happening
    // const app = await ctx.$fire.appReady()
  })

  test('exec plugin (client)', async () => {
    process.client = true
    const Plugin = await import(resolve(buildDir, 'firebase/index.js')).then(m => m.default || m)
    const ctx = {}
    const inject = jest.fn()

    await Plugin(ctx, inject)
    expect(inject).toHaveBeenCalledTimes(1)
    expect(ctx.$fire).toBeDefined()
    expect(ctx.$fireModule).toBeUndefined()
  })
})
