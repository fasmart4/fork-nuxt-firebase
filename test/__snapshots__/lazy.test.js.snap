// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lazy plugin contents 1`] = `
"import createApp from './app.js'

import authService from './service.auth.js'
import databaseService from './service.database.js'
import firestoreService from './service.firestore.js'
import storageService from './service.storage.js'
import functionsService from './service.functions.js'
import messagingService from './service.messaging.js'
import performanceService from './service.performance.js'
import analyticsService from './service.analytics.js'
import remoteConfigService from './service.remoteConfig.js'

export default async (ctx, inject) => {
  // Runtime configs
  const appConfig = ctx.$config && ctx.$config.firebaseConfig || {\\"apiKey\\":\\"\\\\u003CapiKey\\\\u003E\\",\\"authDomain\\":\\"\\\\u003CauthDomain\\\\u003E\\",\\"databaseURL\\":\\"\\\\u003CdatabaseURL\\\\u003E\\",\\"projectId\\":\\"\\\\u003CprojectId\\\\u003E\\",\\"storageBucket\\":\\"\\\\u003CstorageBucket\\\\u003E\\",\\"messagingSenderId\\":\\"\\\\u003CmessagingSenderId\\\\u003E\\",\\"appId\\":\\"\\\\u003CappId\\\\u003E\\",\\"measurementId\\":\\"\\\\u003CmeasurementId\\\\u003E\\"}

  let firebase, session
  let firebaseReady = false

  const fire = {
    async appReady() {
      if (!firebaseReady) {
        ({ firebase, session } = await createApp(appConfig, ctx))
        firebaseReady = true;
      }
      return session
    },
    async ready() {
      await fire.appReady()

      let servicePromises = []

      if (process.server) {
        servicePromises = [
        fire.authReady(),
        fire.databaseReady(),
        fire.firestoreReady(),
        fire.storageReady(),
        fire.functionsReady(),

        ]
      }

      if (process.client) {
        servicePromises = [
        fire.authReady(),
        fire.databaseReady(),
        fire.firestoreReady(),
        fire.storageReady(),
        fire.functionsReady(),
        fire.messagingReady(),
        fire.performanceReady(),
        fire.analyticsReady(),
        fire.remoteConfigReady(),

        ]
      }

      await Promise.all(servicePromises)
      return session
    }
  }

  if (process.server) {
  fire.auth = null
  fire.authReady = async () => {
    if (!fire.auth) {
      await fire.appReady()
      fire.auth = await authService(session, firebase, ctx, inject)
    }

    return fire.auth
  }

  fire.database = null
  fire.databaseReady = async () => {
    if (!fire.database) {
      await fire.appReady()
      fire.database = await databaseService(session, firebase, ctx, inject)
    }

    return fire.database
  }

  fire.firestore = null
  fire.firestoreReady = async () => {
    if (!fire.firestore) {
      await fire.appReady()
      fire.firestore = await firestoreService(session, firebase, ctx, inject)
    }

    return fire.firestore
  }

  fire.storage = null
  fire.storageReady = async () => {
    if (!fire.storage) {
      await fire.appReady()
      fire.storage = await storageService(session, firebase, ctx, inject)
    }

    return fire.storage
  }

  fire.functions = null
  fire.functionsReady = async () => {
    if (!fire.functions) {
      await fire.appReady()
      fire.functions = await functionsService(session, firebase, ctx, inject)
    }

    return fire.functions
  }
  }

  if (process.client) {
    fire.auth = null
    fire.authReady = async () => {
      if (!fire.auth) {
        await fire.appReady()
        fire.auth = await authService(session, firebase, ctx, inject)
      }

      return fire.auth
    }

    fire.database = null
    fire.databaseReady = async () => {
      if (!fire.database) {
        await fire.appReady()
        fire.database = await databaseService(session, firebase, ctx, inject)
      }

      return fire.database
    }

    fire.firestore = null
    fire.firestoreReady = async () => {
      if (!fire.firestore) {
        await fire.appReady()
        fire.firestore = await firestoreService(session, firebase, ctx, inject)
      }

      return fire.firestore
    }

    fire.storage = null
    fire.storageReady = async () => {
      if (!fire.storage) {
        await fire.appReady()
        fire.storage = await storageService(session, firebase, ctx, inject)
      }

      return fire.storage
    }

    fire.functions = null
    fire.functionsReady = async () => {
      if (!fire.functions) {
        await fire.appReady()
        fire.functions = await functionsService(session, firebase, ctx, inject)
      }

      return fire.functions
    }

    fire.messaging = null
    fire.messagingReady = async () => {
      if (!fire.messaging) {
        await fire.appReady()
        fire.messaging = await messagingService(session, firebase, ctx, inject)
      }

      return fire.messaging
    }

    fire.performance = null
    fire.performanceReady = async () => {
      if (!fire.performance) {
        await fire.appReady()
        fire.performance = await performanceService(session, firebase, ctx, inject)
      }

      return fire.performance
    }

    fire.analytics = null
    fire.analyticsReady = async () => {
      if (!fire.analytics) {
        await fire.appReady()
        fire.analytics = await analyticsService(session, firebase, ctx, inject)
      }

      return fire.analytics
    }

    fire.remoteConfig = null
    fire.remoteConfigReady = async () => {
      if (!fire.remoteConfig) {
        await fire.appReady()
        fire.remoteConfig = await remoteConfigService(session, firebase, ctx, inject)
      }

      return fire.remoteConfig
    }
  }

  inject('fire', fire)
  ctx.$fire = fire
}

function forceInject (ctx, inject, key, value) {
  inject(key, value)
  const injectKey = '$' + key
  ctx[injectKey] = value
  if (typeof window !== \\"undefined\\" && window.$nuxt) {
  // If clause makes sure it's only run when ready() is called in a component, not in a plugin.
    window.$nuxt.$options[injectKey] = value
  }
}"
`;
